### PR TITLE
Add reponse headers property

### PR DESCRIPTION
**Pull request by ganymedes01**
It would look something like this and would allow us to access the header properties that are declared in the openapi spec
```ts
{
    readonly status: 201;
    readonly value: {
        readonly contentType: "application/json";
        readonly value: common.ShapeOfItem;
    };
    readonly headers: {
        readonly location: string;
    };
}
```

https://github.com/smartlyio/oats-runtime/pull/106
https://github.com/smartlyio/oats-axios-adapter/pull/42

### CODE REVIEW

File: src/generate-types.ts [hunk 1]
```diff
@@ -352,6 +352,26 @@ export function run(options: Options) {
...
+        required.push(headerName);
+        return { ...memo, [headerName]: headerObject.schema };
+      }
+      return memo;
+    }, {} as { [key: string]: oas.SchemaObject | oas.ReferenceObject });
+    return {
+      type: 'object',
+      properties,
+      required,
+      additionalProperties: true
```
Comment (ganymedes01) [inline src/generate-types.ts:371]: `true` so that we don't need to declare *all*  the headers

File: src/generate-types.ts [hunk 2]
```diff
@@ -352,6 +352,26 @@ export function run(options: Options) {
...
+      } else if (headerObject.schema) {
+        required.push(headerName);
+        return { ...memo, [headerName]: headerObject.schema };
+      }
+      return memo;
+    }, {} as { [key: string]: oas.SchemaObject | oas.ReferenceObject });
+    return {
+      type: 'object',
+      properties,
+      required,
```
Comment (ganymedes01) [inline src/generate-types.ts:370]: should we make the declared headers `required`? 🤔

Comment (ganymedes01) [reply src/generate-types.ts:370]: The specification doesn't mention it: https://swagger.io/docs/specification/describing-responses/

Comment (sakari) [reply src/generate-types.ts:370]: There is a required flag you can give to the header parameter. If the thing is required then yes otherwise I think not.

https://swagger.io/docs/specification/describing-parameters/#header-parameters

File: src/generate-types.ts [hunk 3]
```diff
@@ -423,6 +443,10 @@ export function run(options: Options) {
           required: ['status', 'value'],
           additionalProperties: false
         };
+        if (!oautil.isReferenceObject(response) && response.headers) {
+          schema.properties!.headers = generateHeadersSchemaType(response.headers);
+          schema.required!.push('headers');
```
Comment (sakari) [inline src/generate-types.ts:448]: so to check we would only need to provide a headers object in the response if there are response headers defined in the spec?

**Comment by sakari** [comment]
you could add an test for this by copy pasting and changing test/optional-queries/* to suit this use case. The gist there is to make it somewhat easy to test various api knobs.

**Comment by sakari** [comment]
tldr; sweet, was just wondering whether I should do this a few days ago

File: src/generate-types.ts (hunk 2 above)
Comment (ganymedes01) [reply src/generate-types.ts:370]: Hm, what about `$ref` headers? is there an easy way to "resolve" the ref so we can check if that's required?

File: src/generate-types.ts (hunk 3 above)
Comment (ganymedes01) [reply src/generate-types.ts:448]: I guess I could also check if there's actually any headers declared in there, so we don't end up having to set empty objects as `headers` 👍

File: src/generate-types.ts (hunk 2 above)
Comment (sakari) [reply src/generate-types.ts:370]: Point. not really as such if we consider that the ref can be to an another file. Best we could do is emit the equivalent of `undefined | parameterType` for parameter components, but it would still require the property to be there. Unless we want to have a look on the runtime type checking code for undefined values.

making $ref optional might be ok or even making everything optional

Comment (sakari) [reply src/generate-types.ts:370]: Copoint. It actually might be that we inline parameter schemas. so this should actually work out of the box (mostly bc inlining would fail for out of file references)

yes. we do if you look eg at generateParameterType in generate-types.ts it uses deref which does this.

**Review by sakari — APPROVED** [review]
nice. with or without tests lgtm.

File: src/generate-types.ts (hunk 1 above)
Comment (sakari) [reply src/generate-types.ts:371]: we drop extra headers for requests in server side would that work here or are we keen to allow also unknown headers?

Comment (sakari) [reply src/generate-types.ts:371]: additionalproperties are a bit me in general as those get turned into `{[key]: unknown}` which are not really nice to use.

**Review by sakari — CHANGES_REQUESTED** [review]
maybe rethink a bit whether allowing additionalProperties for response headers is nice

**Comment by sakari** [comment]
maybe typing the additional headers as strings would be nicer as that is pretty much what those are going to end up being

**Review by sakari — APPROVED** [review]

**Comment by ganymedes01** [comment]
I agree that not allowing extra headers might be ideal.
how would we go about that? I guess it would be the adapter handling the request's job to drop the extra fields, but afaik it has no knowledge of which fields it should keep/omit

I guess it could change the oats-runtime to make using that `{ unknownField: 'drop' }` option, but I think that would require a bit of refactoring
